Furthest points

erickzul opened the pull request:
Adds FURTHEST_POINTS partition strategy.
  What this strategy consists of is, attempting max_attempts times:

  1. Get top_n coordinates from the sample that are the furthest away
     from the centroid of the overall sample.
  2. Sample randomly 3 of these top_n coordinates. This will be the
     centroid of the train/validation/test splits.
  3. Get the sample_size*fraction nearest neighbors to each centroid.
  4. Check that new splits don't overlap or have an area smaller than 0.1

Experiment that resulted in choosing this strategy:
https://docs.google.com/document/d/1AWdJZXn9sC8KZ3ggsG7PM9BwJsP357kxX6sHVH3DNBI/edit?usp=sharing

erickzul commented:
Offline discussion notes and answers based on @rothn 's notes and questions:
https://docs.google.com/document/d/1Jo5Ey5t3rhJkFMHLZQBADc78Pl3leKx_IHL0uq9ny9k/edit?usp=sharing

rothn commented on dataset.py:
```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
+    n_neighbors=sample_data.shape[0]*strategy.validation_fraction 
+  )
+  filtered_sample_data = filtered_sample_data[
+    ~filtered_sample_data.index.isin(
```
This is a bitwise operator. I think we probably want the logical NOT operator `not`

erickzul replied:
Within pandas dataframe operations, ~ is the required operator. NOT is not compatible.

See:
https://screenshot.googleplex.com/AThFnL3WAKNK2nU
https://sparkbyexamples.com/pandas/not-in-filter-in-pandas/

rothn commented on dataset.py:
```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
+    shuffled_coordinates_df[
+      [_LONGITUDE_COLUMN_NAME, _LATITUDE_COLUMN_NAME]
+    ].values)
+  assert(len(coordinates) >= strategy.top_n)
```
These asserts should probably be some kind of an issue-specific exception. I generally don't see this in libraries. Here's one stackoverflow post that discusses this: https://stackoverflow.com/questions/3674962/assert-vs-exception-when-developing-the-library

Exceptions can provide more information about what's going wrong and it's just more of a standard way to validate API parameters. I guess asserts here could make sense if the external-facing part of this API has no chance of triggering them unless we did something wrong. Is that the case here?

rothn replied:
Our internal style guidelines are even more prescriptive than this, but we have a good amount of flexibility here since this is external code.

rothn commented on dataset.py:
```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
+  min_size = (int(len(coordinates) * strategy.train_fraction) +
+    int(len(coordinates) * strategy.validation_fraction) +
+    int(len(coordinates) * strategy.test_fraction))
+  assert(int(len(coordinates) * strategy.train_fraction) > 0)
+  assert(int(len(coordinates) * strategy.validation_fraction) > 0)
+  assert(int(len(coordinates) * strategy.test_fraction) > 0)
```
Do we check this somewhere else and potentially raise an exception?

rothn commented on dataset.py:
```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
+  assert(int(len(coordinates) * strategy.train_fraction) > 0)
+  assert(int(len(coordinates) * strategy.validation_fraction) > 0)
+  assert(int(len(coordinates) * strategy.test_fraction) > 0)
+  assert(
+    min_size <= len(coordinates)
+  ), (f"You need {min_size} aka {int(len(coordinates) * strategy.train_fraction)} train + "
+    f"{int(len(coordinates) * strategy.validation_fraction)} validation + "
+    f"{int(len(coordinates) * strategy.test_fraction)} test samples but you have a sample of {len(coordinates)}")
```
If this is meant to be externally-facing, should be a ValueError

rothn commented on dataset.py:
```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
+  Returns a PartitionedDataset that has a valid train/validation/test split
+  that has:
+  - Non-overlapping splits
+  - Splits which polygon areas have less than 0.1 area
```
Can you clarify what "Splits which polygon areas have less than 0.1 area" means?

erickzul replied:
Ack, done.

rothn commented on dataset.py:
```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
+  '''
+  partitioned_dataset = None
+  attempts = 0
+  while (attempts <= strategy.max_attempts and
+         partitioned_dataset is None):
```
nit: A more Pythonic way to do this might be `for attempt in range(strategy.max_attempts):.... break`. When I first came to Python from C++, I would usually do this instead so in a way I'm just passing on feedback from others :-).

Either way, I think people will understand what you mean. But in the Python community, we tend to be faster at reading the `for` loop variant of this.

erickzul replied:
Sounds good. Updated.

rothn commented on dataset.py:
```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
+
+def _maybe_partition_furthest_points(
+  sample_data: pd.DataFrame,
+  furthest_coordinates: list[list[float]],
```
Can you add an Args: block in the comments that explain what the list[list[float]] is for?

I think maybe this is supposed to be a `float **furthest_coordinates` which would make a lot of sense in C but in Python we would prefer a `List[Tuple[float, float]]`.

Side-note: Does your C++ team do this? Because in C++ I would usually do a `std::vector<std::pair<float, float>>` or `std::vector<std::tuple<float, float>>` or maybe even create a type and do `std::vector<CoordinatePair>` if the meaning of the former was unclear.

erickzul replied:
Created Coordinate class.

rothn commented on dataset.py:
```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
+  distances_to_centroid = []
+  for coord in coordinates:
+    distances_to_centroid.append((
+      math.dist(centroid, coord), (coord[0], coord[1])))
```
It might be good to specify in the externally-facing docstring that the specific distance we use is Euclidean.

erickzul replied:
Done.

rothn reviewed (approved):
This is really cool. Thanks for figuring this out and running all of those experiments!

I've got a few code comments here, mostly focused on readability, maintainability, and external usage. Nothing that should take too long to fix, though!

rothn reviewed (changes requested):
To be clear, I'd like some of this stuff I mentioned to be addressed :-).

erickzul commented on dataset.py:
```diff
@@ -1,18 +1,22 @@
 import random
-import datetime
+import raster
 
```
I'm just separating the "from x" imports for the regular "import x" imports.

rothn replied:
Okay. I am much more familiar with the "skip a line between libraries and imported dependencies" convention, and that would be my preference, per pystyle. Let me know if you have a good reason for doing it this way.

rothn replied:
(bump)

erickzul commented on dataset.py:
```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
+  From sample_data, select the n_neighbors closest to the specified
+  center_coordinate.
+  '''
+  assert(len(center_coordinate) == 2)
```
Ack, done.

erickzul commented on dataset.py:
```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
+    n_neighbors=sample_data.shape[0]*strategy.train_fraction
+  )
+  filtered_sample_data = sample_data[
+          ~sample_data.index.isin(
```
Within pandas dataframe operations, ~ is the required operator. NOT is not compatible.
See:
https://screenshot.googleplex.com/AThFnL3WAKNK2nU
https://sparkbyexamples.com/pandas/not-in-filter-in-pandas/

erickzul commented on dataset.py:
```diff
@@ -91,6 +95,45 @@ class RandomPartitionStrategy:
   0.8, 0.1, 0.1, None
 )
 
+@dataclass
```
Could you specify the lines of this block?

rothn replied:
Lines 100-433

erickzul commented:
Moved partition method and its helper functions and classes to partitioned_dataset.py
Split tests this way as well.

Rearranged imports but lmk if there's a different way to do it.

rothn reviewed (changes requested):
I really like how the imports are more in line with what we expect to see in pystyle and PEP-8 (link: https://peps.python.org/pep-0008/#imports) more generally.

However, I'm curious about the intention behind the refactor. My suggestion was to move the code that supports the new partitioning strategy into its own file since it is large. I'm trying to understand the rationale for the way this change was made.

One thing that confuses me is that I can see what appear to be an existing in-use public APIs in `dataset` split into a new file. Is this the case?
